### PR TITLE
200 kafkaW producer/consumer interface

### LINKED ISSUE #200: Add interface for the producer in KafkaW
This means we can plug in a fake producer for testing. currently the consumer has one so can be used as reference

### LINKED ISSUE #207: Create ConsumerInterface in kafkaW
create an interface so that it's possible to mock Consumer

### PR DESCRIPTION
**Pull request by mchorazak**
### Description of work

introduced both Producer and Consumer interface.
Variables' names changed to make the code more readable.

### Issue

Closes #200 
Closes #207 

### Acceptance Criteria

Producer.h/cpp
Consumer.h/cpp

---

## Code Review (To be filled in by the reviewer only)

- [x] Is the code of an acceptable quality?
- [x] Do the changes function as described and is it robust?

---

## Nominate for Group Code Review (Anyone can nominate it)
Indicate if you think the code should be reviewed in a Thursday code review session.

- [ ] Recommend for group code review

Also, nominate it on the code_review Slack channel (does someone want to automate this?).

### CODE REVIEW

File: src/KafkaW/Producer.cpp [hunk 1]
```diff
@@ -9,104 +9,104 @@ void ProducerMsg::deliveryOk() {}
...
+      if (EventsHandled > 0) {
         LOG(Sev::Debug,
             "rd_kafka_poll handled: {}  outq before: {}  timeout: {}",
-            events_handled, outq_len, timeout_ms);
+            EventsHandled, OutQueueLength, Timeout_ms);
       }
-      timeout_ms = timeout_ms << 1;
-      if (timeout_ms > 8 * 1024) {
+      Timeout_ms = Timeout_ms << 1;
+      if (Timeout_ms > 8 * 1024) {
```
Comment (rerpha) [inline src/KafkaW/Producer.cpp]: could simplify this

File: src/KafkaW/Producer.cpp [hunk 2]
```diff
@@ -9,104 +9,104 @@ void ProducerMsg::deliveryOk() {}
...
-      self->id, broker_id, broker_name, throttle_time_ms);
+      Self->id, Broker_id, Name, Throttle_time_ms);
 }
 
 Producer::~Producer() {
   LOG(Sev::Debug, "~Producer");
   if (RdKafkaPtr) {
-    int timeout_ms = 1;
-    uint32_t outq_len = 0;
+    int Timeout_ms = 1;
```
Comment (rerpha) [inline src/KafkaW/Producer.cpp]: TimeoutMS

File: src/KafkaW/Producer.h [hunk 3]
```diff
@@ -33,69 +33,81 @@ struct ProducerStats {
...
-class Producer {
+class ProducerInterface {
+public:
+  ProducerInterface() = default;
+  virtual ~ProducerInterface() = default;
+  virtual void pollWhileOutputQueueFilled() = 0;
+  virtual void poll() = 0;
+  virtual uint64_t totalMessagesProduced() = 0;
+  virtual uint64_t outputQueueLength() = 0;
+  virtual rd_kafka_t *getRdKafkaPtr() const = 0;
```
Comment (rerpha) [inline src/KafkaW/Producer.h]: looks good :+1:

File: src/KafkaW/Producer.h [hunk 4]
```diff
@@ -33,69 +33,81 @@ struct ProducerStats {
...
   typedef ProducerTopic Topic;
   typedef ProducerMsg Msg;
   explicit Producer(BrokerSettings ProducerBrokerSettings_);
   Producer(Producer const &) = delete;
   Producer(Producer &&x) noexcept;
-  ~Producer();
-  void pollWhileOutputQueueFilled();
+  ~Producer() override;
+  void pollWhileOutputQueueFilled() override;
   void poll();
```
Comment (rerpha) [inline src/KafkaW/Producer.h]: does this need an override?

File: src/KafkaW/Producer.cpp [hunk 5]
```diff
@@ -164,15 +164,15 @@ Producer::Producer(Producer &&x) noexcept {
...
 
 void Producer::poll() {
-  int events_handled =
+  int EventsHandled =
       rd_kafka_poll(RdKafkaPtr, ProducerBrokerSettings.PollTimeoutMS);
   LOG(Sev::Debug,
       "IID: {}  broker: {}  rd_kafka_poll()  served: {}  outq_len: {}", id,
-      ProducerBrokerSettings.Address, events_handled, outputQueueLength());
+      ProducerBrokerSettings.Address, EventsHandled, outputQueueLength());
   if (log_level >= 8) {
```
Comment (rerpha) [inline src/KafkaW/Producer.cpp]: thinking about it, we should just remove this. the log level should never be higher than 7 anyway.

**Comment by rerpha** [comment]
to me it looks like `pollWhileOutputQueueFilled()` and `totalMessagesProduced()` are never used in the forwarder. I think if this is the case they should be removed from the main code and the interface.

**Comment by SkyToGround** [comment]
@rerpha I am all for removing dead code.

**Pull request merged**